### PR TITLE
Adds annotations to secure sensible links against CSRF attacks

### PR DESCRIPTION
**Pull request by Tobsucht**

### CODE REVIEW

File: src/main/java/sirius/web/controller/ControllerDispatcher.java [hunk 1]
```diff
@@ -129,6 +132,12 @@ private boolean tryExecuteRoute(WebContext ctx, boolean preDispatch, String uri,
         }
     }
 
+    private boolean checkCSRFToken(WebContext ctx) {
+        String requestToken = ctx.get(CSRF_TOKEN).asString();
+        return Strings.isFilled(requestToken) && Strings.areEqual(requestToken,
+                                                                  ctx.getServerSession().getValue(CSRF_TOKEN).asString());
```
Comment (andyHa) [inline src/main/java/sirius/web/controller/ControllerDispatcher.java]: nope - not the server session

File: src/main/java/sirius/web/http/session/MemoryServerSession.java [hunk 2]
```diff
@@ -12,9 +12,11 @@
 import com.google.common.collect.Maps;
 import sirius.kernel.commons.Value;
 import sirius.kernel.di.std.ConfigValue;
+import sirius.web.controller.ControllerDispatcher;
```
Comment (andyHa) [inline src/main/java/sirius/web/http/session/MemoryServerSession.java]: client session...

File: src/main/java/sirius/web/controller/CheckSecurityToken.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package sirius.web.controller;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Marks a {@link Routed} method, which modifies sensitiv data so that a secret security-session-token is checked
```
Comment (meg1502) [inline src/main/java/sirius/web/controller/CheckSecurityToken.java]: sensitivE

File: src/main/java/sirius/web/controller/AddSecurityToken.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package sirius.web.controller;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Marks a {@link Routed} method, so that a secret security-session-token should be added to action links which modify
+ * sensitiv data. The security token is automatically placed/send when using the delete-link taglib
```
Comment (meg1502) [inline src/main/java/sirius/web/controller/AddSecurityToken.java]: sensitivE

File: src/main/java/sirius/web/controller/CheckSecurityToken.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ * against the provided POST parameter. Must be combined with another {@link AddSecurityToken} method which adds the
+ * security token to sensitiv action links which call the {@link CheckSecurityToken} annotated method.
```
Comment (meg1502) [inline src/main/java/sirius/web/controller/CheckSecurityToken.java]: sensitivE

File: src/main/java/sirius/web/http/WebContext.java [hunk 6]
```diff
@@ -1078,6 +1091,50 @@ public boolean isSSL() {
         return ssl;
     }
 
+    /**
+     * Returns the CSRF security-token of the current session.
+     *
+     * @return the CSRF security-token to protect sensitv links.
```
Comment (meg1502) [inline src/main/java/sirius/web/http/WebContext.java]: sensitIvE

File: src/main/java/sirius/web/http/WebContext.java [hunk 7]
```diff
@@ -293,6 +301,11 @@
      */
     protected UserAgent userAgent;
 
+    /*
+      * Whether CSRF tokens should be placed into sensitiv links.
```
Comment (meg1502) [inline src/main/java/sirius/web/http/WebContext.java]: sensitivE

File: src/main/java/sirius/web/http/WebContext.java [hunk 8]
```diff
@@ -1078,6 +1091,50 @@ public boolean isSSL() {
...
+        }
+    }
+
+    private boolean isCSRFTokenOutdated(long lastCSRFRecompute) {
+        return Duration.between(Instant.ofEpochMilli(lastCSRFRecompute), Instant.now()).toMinutes()
+               > csrfTokenLifetime.toMinutes();
+    }
+
+    /**
+     * Returns whether CSRF security-tokens should be placed in sensitiv action links.
```
Comment (meg1502) [inline src/main/java/sirius/web/http/WebContext.java]: sensitivE

File: src/main/java/sirius/web/http/WebContext.java [hunk 9]
```diff
@@ -1078,6 +1091,50 @@ public boolean isSSL() {
...
+     *
+     * @return <tt>true</tt> if CSRF security-tokens should be placed in sensitiv action links, <tt>false</tt> otherwise.
```
Comment (meg1502) [inline src/main/java/sirius/web/http/WebContext.java]: sensitivE

File: src/main/java/sirius/web/http/WebContext.java [hunk 10]
```diff
@@ -1078,6 +1091,50 @@ public boolean isSSL() {
...
+     */
+    public boolean isUseCSRFProtection() {
+        return useCSRFProtection;
+    }
+
+    /**
+     * Marks the current {@link WebContext} context so that CSRF security-tokens should be placed in sensitiv action
```
Comment (meg1502) [inline src/main/java/sirius/web/http/WebContext.java]: sensitivE

File: src/main/java/sirius/web/http/session/ServerSession.java [hunk 11]
```diff
@@ -167,4 +167,11 @@
      * session live time over a session without an attached user.
      */
     void markAsUserSession();
+
+    /**
+     * Returns the current CSRF security-token.
+     *
+     * @return the security token which should be placed in sensitiv action links
```
Comment (meg1502) [inline src/main/java/sirius/web/http/session/ServerSession.java]: sensitivE

File: src/main/resources/web_de.properties [hunk 12]
```diff
@@ -89,4 +89,6 @@ ScopeDefaultConfigController.files=Datei
 Filterbox.noItems=Keine Filterkategorien gefunden
 
 fileUpload.specifyURL=URL angeben
-fileUpload.URL=URL
\ No newline at end of file
+fileUpload.URL=URL
+
+ControllerDispatcher.invalidCSRFToken=Abgelaufenes oder invalides Sicherheitstoken. F\u00FChren Sie ihre letzte Aktion bitte erneut aus.
```
Comment (meg1502) [inline src/main/resources/web_de.properties]: If "Sie" is captialized, "ihre" should be capitalized too ?

File: src/main/resources/web_en.properties [hunk 13]
```diff
@@ -84,3 +84,5 @@ template.html.uploadBtn=Choose a file...
 template.html.uploadFailed=Upload failed
 
 Filterbox.noItems=No filter categories found
+
+ControllerDispatcher.invalidCSRFToken=Invalid security token! Please retry your last action.
```
Comment (meg1502) [inline src/main/resources/web_en.properties]: Translation mismatch "Invalid security token" EN <--> DE "Expired or invalid security token" 

german text has more info

File: src/main/resources/component-web.conf [hunk 14]
```diff
@@ -103,6 +103,9 @@ http {
     # session is eligible for invalidation.
     serverUserSessionLifetime = 2 hours
 
+    # Specifies the lifetime of CSRF security tokens in a user session before being recomputed.
+    csrfTokenLifetime = 2 hours
```
Comment (Tobsucht) [inline src/main/resources/component-web.conf]: should we increase this value to ~12hours to not limit user experience?

**Pull request closed without merging**